src/tls/incoming.rs

```diff
@@ -357,3 +359,109 @@ impl AsyncWrite for MaybeTlsIncomingStream<TcpStream> {
         }
     }
 }
+
+pub struct CertificateMetadata {
+    pub country_name: Option<String>,
+    pub state_or_province_name: Option<String>,
+    pub locality_name: Option<String>,
+    pub organization_name: Option<String>,
+    pub organizational_unit_name: Option<String>,
+    pub common_name: Option<String>,
+}
+
+impl CertificateMetadata {
+    pub(crate) fn from_x509(cert: X509) -> CertificateMetadata {
+        let mut subject_metadata: HashMap<String, String> = HashMap::new();
+        for entry in cert.subject_name().entries() {
+            let data_string = match entry.data().as_utf8() {
+                Ok(data) => data.to_string(),
+                Err(_) => "".to_string(),
+            };
+            subject_metadata.insert(entry.object().to_string(), data_string);
+        }
+        return CertificateMetadata {
+            country_name: subject_metadata.get("countryName").cloned(),
+            state_or_province_name: subject_metadata.get("stateOrProvinceName").cloned(),
+            locality_name: subject_metadata.get("localityName").cloned(),
+            organization_name: subject_metadata.get("organizationName").cloned(),
+            organizational_unit_name: subject_metadata.get("organizationalUnitName").cloned(),
+            common_name: subject_metadata.get("commonName").cloned(),
+        }
+    }
+}
+
+impl ToString for CertificateMetadata {
+    fn to_string(&self) -> String {
+        let mut components = Vec::<String>::new();
+        match &self.common_name {
+            Some(cn) => components.push(format!("CN={}", cn)),
+            None => (),
+        }
```

Comment:
Typically you'd do `if let Some(cn) = &self.common_name` here instead of `match`.

Reply:
Since you're doing nothing in the `None` branch, using `if let` is simpler here too:
```suggestion
        if let Some(cn) = &self.common_name {
            components.push(format!("CN={}", cn)),
        }
```